packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1541,72 +1268,336 @@ module.exports = function<T, P, I, TI, PI, C, CC, CX, PL>(
     return mostRecentCurrentTime;
   }
 
+  function deferredUpdates<A>(fn: () => A): A {
+    const previousExpirationContext = expirationContext;
+    expirationContext = computeAsyncExpiration();
+    try {
+      return fn();
+    } finally {
+      expirationContext = previousExpirationContext;
+    }
+  }
+
+  function syncUpdates<A>(fn: () => A): A {
+    const previousExpirationContext = expirationContext;
+    expirationContext = Sync;
+    try {
+      return fn();
+    } finally {
+      expirationContext = previousExpirationContext;
+    }
+  }
+
+  // TODO: Everything below this is written as if it has been lifted to the
+  // renderers. I'll do this in a follow-up.
+
+  // Linked-list of roots
+  let firstScheduledRoot: FiberRoot | null = null;
+  let lastScheduledRoot: FiberRoot | null = null;
+
+  let isCallbackScheduled: boolean = false;
+  let isRendering: boolean = false;
+  let nextFlushedRoot: FiberRoot | null = null;
+  let nextFlushedExpirationTime: ExpirationTime = NoWork;
+  let deadlineDidExpire: boolean = false;
+  let hasUnhandledError: boolean = false;
+  let unhandledError: mixed | null = null;
+  let deadline: Deadline | null = null;
+
+  let isBatchingUpdates: boolean = false;
+  let isUnbatchingUpdates: boolean = false;
+
+  // Use these to prevent an infinite loop of nested updates
+  const NESTED_UPDATE_LIMIT = 1000;
+  let nestedUpdateCount: number = 0;
+
+  const timeHeuristicForUnitOfWork = 1;
+
+  // requestWork is called by the scheduler whenever a root receives an update.
+  // It's up to the renderer to call renderRoot at some point in the future.
+  function requestWork(root: FiberRoot, expirationTime: ExpirationTime) {
+    if (nestedUpdateCount > NESTED_UPDATE_LIMIT) {
```

Comment:
@acdlite @sebmarkbage Can I ask that why we use this name? It seems that this is only related 'root', should we rename it to `requestWorkOnRoot` or something else?